### PR TITLE
Namespace field and sorting hooks

### PR DESCRIPTION
**Pull request by Fernthedev**
- **metadata: Add namespaze field and matching function to HookNameMetadata**
- **Basic sorting, time to rewrite to topological graph**
- **Implement proper topological sort (untested)**
- **Add topological sort tests**

### CODE REVIEW

File: shared/capi.h [hunk 1]
```diff
@@ -249,6 +259,20 @@ FLAMINGO_C_EXPORT FlamingoUninstallResult flamingo_uninstall_hook(FlamingoHookHa
...
 FLAMINGO_C_EXPORT_VOID void flamingo_format_error(FlamingoInstallErrorData* error, char* buffer, size_t buffer_size);
 
+/// @brief Returns the number of hooks installed at `target`. Returns 0 if none or if target is not hooked.
+FLAMINGO_C_EXPORT size_t flamingo_get_hook_count(uint32_t* target);
+
+/// @brief Fills the provided `hooks` array with `FlamingoHookInfo` entries for `target`.
+/// If `capacity` is smaller than the number of hooks, the function returns the required size but does not write
+/// beyond `capacity` elements.
+/// The `name` and `namespaze` fields inside each written `FlamingoHookInfo` are allocated with `malloc` and must
+/// be freed with `flamingo_free_strings` (pass an array of the `name` pointers or `namespaze` pointers respectively).
```
Comment (sc2ad) [inline shared/capi.h]: nit: Here and elsewhere you have flamingo_free_strings mentioned, instead of free_hooks

File: shared/hook-metadata.hpp [hunk 2]
```diff
@@ -24,15 +24,31 @@ struct InstallationMetadata {
 /// Lookups are described using userdata when the HookInfo is made at first.
 struct HookNameMetadata {
   std::string name{};
+  std::string namespaze{};
+
+  /// @brief Checks if this name metadata matches another (either by name or namespace)
+  /// @param other The other metadata to check against
+  /// @return True if either the name or namespace matches
+  [[nodiscard]] bool matches(HookNameMetadata const& other) const {
+    return *this == other || (!name.empty() && name == other.name) || (!namespaze.empty() && namespaze == other.namespaze);
```
Comment (sc2ad) [inline shared/hook-metadata.hpp]: Doing the operator== first makes the rest of the check largely redundant, I also don't follow why namespace matching OR name matching is preferable, it feels like such a concept shouldn't even be exposed (name matching should never happen unless namespace matching happens first, for example)

Also, you can just check the namespace string unconditionally, since it'll need to be empty in both cases to match (even if this function were to exist)

File: shared/hook-metadata.hpp [hunk 3]
```diff
@@ -60,6 +76,6 @@ class fmt::formatter<flamingo::HookNameMetadata> {
   }
   template <typename Context>
   constexpr auto format(flamingo::HookNameMetadata const& metadata, Context& ctx) const {
-    return fmt::format_to(ctx.out(), "name: {}", metadata.name);
+    return fmt::format_to(ctx.out(), "name: {} namespaze {}", metadata.name, metadata.namespaze);
```
Comment (sc2ad) [inline shared/hook-metadata.hpp]: nit: string can read namespace

File: src/capi.cpp [hunk 4]
```diff
@@ -98,6 +100,9 @@ FlamingoUninstallResult convert_uninstall_result(flamingo::Result<bool, bool> co
 FLAMINGO_C_EXPORT FlamingoNameInfo* flamingo_make_name(char const* name_str) {
   return reinterpret_cast<FlamingoNameInfo*>(new flamingo::HookNameMetadata{ .name = name_str });
 }
+FLAMINGO_C_EXPORT FlamingoNameInfo* flamingo_make_name_namespaced(char const* name_str, char const* namespaze_str) {
```
Comment (sc2ad) [inline src/capi.cpp]: I feel like it generally makes more sense to have namespace first, unless the goal is to have a default parameter here or something

File: src/fixups.cpp [hunk 5]
```diff
@@ -540,6 +540,10 @@ void ShimTarget::WriteJump(void* address) {
   // The writer for ensuring correct permissions and also performing the write
   ProtectionWriter<uint32_t> writer(*this);
   WriteCallback(writer, reinterpret_cast<uint32_t*>(address));
+
+  // flush instruction cache
+  __builtin___clear_cache(reinterpret_cast<char*>(addr.data()),
```
Comment (sc2ad) [inline src/fixups.cpp]: Imo, I think this should be moved to the protection writer's dtor, since we need to flush icache if we touch instructions ever, but only once when we kill the instance to "finish" our writes of a particular region

File: test/test-wrapper.hpp [hunk 6]
```diff
@@ -29,10 +29,12 @@
 // - Dump the trampoline
 // - Dump the new hook
 // Perhaps we make TestWrapper take a Fixups instance for this? That way we can do all three of those.
-#define ERROR(S, ...)                              \
-  fmt::print(stderr, FMT_COMPILE(S), __VA_ARGS__); \
-  fmt::print(stderr, "\n");                        \
-  std::exit(1);
+#define ERROR(S, ...)                     \
+  do {                                    \
+    fmt::print(stderr, S, ##__VA_ARGS__); \
```
Comment (sc2ad) [inline test/test-wrapper.hpp]: Why ##va_args here? This will stirngify any input parameter (ex, x --> "x" instead of being formatted)

File: test.sh [hunk 7]
```diff
@@ -0,0 +1,5 @@
+#!/bin/bash
```
Comment (sc2ad) [inline test.sh:1]: Can you also run this in CI? Maybe instead of the current tests that are done that mirror the logic?

**Review by sc2ad — COMMENTED** [review]
Haven't reviewed the sorting stuff yet but I'll comment with this

File: shared/hook-metadata.hpp (hunk 2 above)
Comment (Fernthedev) [reply shared/hook-metadata.hpp]: I think this was largely done for backwards compatibility (since this was a semi-breaking change). However I can just move the PR to a namespaze first approach and disregard any compatibility or notion that a namespaze isn't provided

Comment (Fernthedev) [reply shared/hook-metadata.hpp]: ok nevermind I rememeber now. I was planning on abusing this struct as a filter for either namespaze or name, where empty values mean null.
e.g filter by namespaze `"foo"|any name` or `any namespaze|name "bar"`

though this is arguably not a good way to do it

File: shared/hook-metadata.hpp (hunk 3 above)
Comment (Fernthedev) [reply shared/hook-metadata.hpp]: did not quite get that

Comment (sc2ad) [reply shared/hook-metadata.hpp]: Could write "namespace: " instead of "namespace: "

File: shared/hook-metadata.hpp [hunk 8]
```diff
@@ -26,21 +26,36 @@ struct HookNameMetadata {
...
+struct HookNameFilter {
+  std::optional<std::string> namespaze{};
+  std::optional<std::string> name{};
+
+  explicit HookNameFilter() = default;
+  explicit HookNameFilter(std::string namespaze) : namespaze(std::move(namespaze)) {}
+  explicit HookNameFilter(std::string namespaze, std::string name) : namespaze(std::move(namespaze)), name(std::move(name)) {}
+
+  /// @brief Construct a filter that matches the provided metadata. This is used for constructing filters from userdata.
+  explicit HookNameFilter(HookNameMetadata const& metadata) : namespaze(metadata.namespaze), name(metadata.name) {}
```
Comment (sc2ad) [inline shared/hook-metadata.hpp]: I think this should actually be an implicit ctor, so we can keep implementation pretty similar

File: .github/workflows/build-ndk.yml [hunk 9]
```diff
@@ -42,17 +42,22 @@ jobs:
...
+      - name: Run test script
         run: |
-          CC=clang CXX=clang cmake -B build -DTEST_BUILD=1 -GNinja
+          cd ${GITHUB_WORKSPACE}
+          ./test.sh
 
-      - name: Build
-        run: |
-          ninja -C build
+      # - name: CMake with LOCAL_TEST enabled
```
Comment (sc2ad) [inline .github/workflows/build-ndk.yml]: nit: I'd just remove the commented out stuff here, no need to keep it

File: shared/fixups.hpp [hunk 10]
```diff
@@ -37,6 +37,10 @@ struct ProtectionWriter {
   ~ProtectionWriter() {
     target.protection = original_permissions;
     target.Protect();
+
+    // flush instruction cache on finish
+    auto addr = target.addr;
+    __builtin___clear_cache(reinterpret_cast<char*>(addr.data()), reinterpret_cast<char*>(addr.data() + addr.size()));
```
Comment (sc2ad) [inline shared/fixups.hpp:40]: This needs to be done before protection is reapplied (since the read permission might be lost)

File: shared/hook-metadata.hpp [hunk 11]
```diff
@@ -60,6 +92,17 @@ class fmt::formatter<flamingo::HookNameMetadata> {
   }
   template <typename Context>
   constexpr auto format(flamingo::HookNameMetadata const& metadata, Context& ctx) const {
-    return fmt::format_to(ctx.out(), "name: {}", metadata.name);
+    return fmt::format_to(ctx.out(), "name: {} namespaze {}", metadata.name, metadata.namespaze);
```
Comment (sc2ad) [inline shared/hook-metadata.hpp]: This string still reads namespaze instead of namespace

File: src/fixups.cpp [hunk 12]
```diff
@@ -540,6 +540,8 @@ void ShimTarget::WriteJump(void* address) {
   // The writer for ensuring correct permissions and also performing the write
   ProtectionWriter<uint32_t> writer(*this);
   WriteCallback(writer, reinterpret_cast<uint32_t*>(address));
+
```
Comment (sc2ad) [inline src/fixups.cpp]: nit: empty space

File: src/installer.cpp [hunk 13]
```diff
@@ -25,8 +31,234 @@ using namespace flamingo;
...
+  }
+
+  // now build topological graph
+  // each hook has a requirement to be after certain other hooks in this graph
+  // we also convert before requirements to after requirements for easier processing
+
+  // build name to iterator map
+  std::unordered_map<HookNameMetadata, std::list<HookInfo>::iterator, HookNameMetadataHash> name_to_iterator;
+  for (auto it = hooks.begin(); it != hooks.end(); ++it) {
+    name_to_iterator[it->metadata.name_info] = it;
```
Comment (sc2ad) [inline src/installer.cpp]: This can be merged into the for loop above

File: src/installer.cpp [hunk 14]
```diff
@@ -25,8 +31,234 @@ using namespace flamingo;
...
+  auto findMatches = [&](HookNameFilter const& filter, HookNameMetadata self) {
+    std::vector<HookNameMetadata> matches;
+    matches.reserve(1);
+    for (auto const& hook : hooks) {
+      auto const& name = hook.metadata.name_info;
+      // exclude self from matches to avoid self-cycle issues. 
+      // If a hook specifies that it should be before/after itself, 
+      // it's likely a mistake, but we won't consider it a cycle since it's not really a dependency.
+      if (name == self) {
+        continue;
```
Comment (sc2ad) [inline src/installer.cpp]: Imo, this SHOULD be an error, it should report a bad priorities installation error eventually. I think the way we can do this is by just checking the incoming hook's before/after set on install instead of checking it during sorting, since we will always only sort well installed hooks.

That'll let you simplify some of the logic here.

File: src/installer.cpp [hunk 15]
```diff
@@ -25,8 +31,234 @@ using namespace flamingo;
...
+  }
+
+  while (!zero_in_degree.empty()) {
+    auto current_name = zero_in_degree.front();
+    zero_in_degree.pop();
+
+    // find the iterator for this name
+    auto it = name_to_iterator.find(current_name);
+    if (it == name_to_iterator.end()) {
+      // should not happen
```
Comment (sc2ad) [inline src/installer.cpp]: I think this should be fatal, not silent

File: src/installer.cpp [hunk 16]
```diff
@@ -25,8 +31,234 @@ using namespace flamingo;
...
+      hook_names.push_back(hook.metadata.name_info.name);
+    }
+    FLAMINGO_DEBUG("Flattened hook order after topological sort attempt: {}", fmt::join(hook_names, " -> "));
+  }
+
+  // now, any remaining hooks in `hooks` are part of cycles
+  // append them in their original order and log a warning. Splicing invalidates
+  // iterators, so consume from the front until empty to preserve original order.
+  for (auto const& hook : hooks) {
+    FLAMINGO_CRITICAL(
```
Comment (sc2ad) [inline src/installer.cpp]: This should be an error or warning log, not critical (since flamingo is not dying, the user tried to install a cyclic dependency)

File: src/installer.cpp [hunk 17]
```diff
@@ -25,8 +31,234 @@ using namespace flamingo;
...
+  // TODO: Do we need to copy Reinstall logic here?
+
+  // Reinstall the orig by calling PerformFixupsAndCallback() again (as needed)
+  // Perform the write of the jump to the first hook
+  // Head
+  auto it = hooks.begin();
+  target_entry->second.fixups.target.WriteJump(it->hook_ptr);
+
+  // Single hook: target -> hook, hook.orig -> fixups or no_fixups
+  if (std::next(it) == hooks.end()) {
```
Comment (sc2ad) [inline src/installer.cpp]: The for loop logic below can be changed to make this logic redundant:

while (std::next(it) != End)
  Assign orig(next(it))
  ++It
It->assign orig(fixups or no fixups)

File: src/installer.cpp [hunk 18]
```diff
@@ -35,20 +267,117 @@ Result<std::list<HookInfo>::iterator, installation::TargetBadPriorities> find_su
...
+      }
+    }
+    // if existing_hook requests to be before us, we cannot install before it
+    for (auto const& before_filter : existing_hook.metadata.priority.befores) {
+      if (before_filter.matches(hook_to_install.metadata.name_info)) {
+        requires_sort = true;
+        break;
+      }
+    }
+    if (requires_sort) {
```
Comment (sc2ad) [inline src/installer.cpp:233]: It might be cleaner to implement this logic in target metadata where when we sort for a given install, we track what names we haven't yet matched, and then when we install a new hook we just check to see if our incoming is in that set to see if we should sort, but that I'll leave as a decision to you if u want

File: src/installer.cpp [hunk 19]
```diff
@@ -35,20 +267,117 @@ Result<std::list<HookInfo>::iterator, installation::TargetBadPriorities> find_su
...
+  }
+
+  // fast track If the incoming hook has no explicit constraints, insert at the front
+  // so newer installs are called before earlier ones (preserve expected install semantics).
+  if (hook_to_install.metadata.priority.afters.empty() && hook_to_install.metadata.priority.befores.empty()) {
+    auto newIt = hooks.emplace(hooks.begin(), std::move(hook_to_install));
+    return ResultT::Ok(newIt);
+  }
+
+  // if no priority constraints affect us, we can install at the first suitable location that fits
```
Comment (sc2ad) [inline src/installer.cpp:276]: I think this logic ends up being slower? I think we should actually think a bit more about this stuff since the ultimate goal should be to prevent recompiles, but the sort could unconditionally happen if we can guarantee some measure of reasonable speed.

It's hard to know if you are a dependee of an existing hook without more state so it can be tricky to track how you should be placed in such a way that makes sense. Also, the linear scan for a given node that has priorities is not trivial, since other hooks might need to be moved around to make space.

The only slow thing is the first node, since swapping the orig call for that is a tiny bit more annoying, but every other node getting reordered should be a trivial pointer chase operation.

I think if we just:
- always sort
- simplify some of the sorting logic to avoid some redundant loops and checks
- ensure edge cases all play nice

We can end up with a simpler impl that also works well. We should discuss this more on discord I think

**Review by sc2ad — COMMENTED** [review]
I haven't read through all the tests, but did finally review the sorting logic. I think we should discuss how to improve it on discord

File: src/installer.cpp [hunk 20]
```diff
@@ -25,8 +31,257 @@ using namespace flamingo;
...
+  for (auto it = hooks.begin(); it != hooks.end(); ++it) {
+    if (it->metadata.priority.is_final) {
+      finals.push_back(it);
+    }
+    name_to_iterator[it->metadata.name_info] = it;
+
+    // check if any reference self in after/before
+    for (auto const& after : it->metadata.priority.afters) {
+      if (after.matches(it->metadata.name_info)) {
+        FLAMINGO_CRITICAL("Hook {} references itself in after dependencies. This is likely a mistake.",
```
Comment (sc2ad) [inline src/installer.cpp]: These should not be CRITICAL logs, since that implies that flamingo itself will crash/fail, this is just user error so should log as warning and return the error

File: src/installer.cpp [hunk 21]
```diff
@@ -25,8 +31,257 @@ using namespace flamingo;
...
+  hooks.swap(sorted_hooks);
+
+  {
+    std::vector<std::string_view> hook_names;
+    hook_names.reserve(hooks.size());
+    for (auto const& hook : hooks) {
+      hook_names.push_back(hook.metadata.name_info.name);
+    }
+    FLAMINGO_DEBUG("Final hook order after topological sort: {}", fmt::join(hook_names, " -> "));
+  }
```
Comment (sc2ad) [inline src/installer.cpp]: Make this a method instead of calling it several times (probably would be useful for generic debugging too). Ideally, we also condition it behind the flamingo debug ifdef, but that doesn't need to be done in this PR

File: src/installer.cpp [hunk 22]
```diff
@@ -25,8 +31,257 @@ using namespace flamingo;
...
+struct HookNameMetadataHash {
+  std::size_t operator()(HookNameMetadata const& k) const {
+    return std::hash<std::string>()(k.name) ^ (std::hash<std::string>()(k.namespaze) << 1);
+  }
+};
+
+/// @brief Topologically sorts the provided hooks by their priority constraints.
+/// @param hooks The list of hooks to sort in place.
+/// @return The sorted list of hooks that have cycles.
+Result<std::list<HookInfo>, installation::TargetBadPriorities> topological_sort_hooks_by_priority(
```
Comment (sc2ad) [inline src/installer.cpp]: I think we still should discuss how we can improve this (and its caller) since it seems quite verbose.

The checking logic for each individual hook also seems verbose (we can instead check when installing a hook whether it has cyclic befores or afters of itself).

The overall set of names --> hooks (ex, the actual topo graph) seems like something we could save per target, since the graph only needs to have things added or removed from it as we install/uninstall hooks.

With that, we could simplify this logic further by basically:
- Check incoming hook for immediate validity
- Add incoming hook and dependencies to graph
- Perform full topo sort with graph
- Update graph iterators to point to new node locations after sorting is complete
- Failures are reported as-is

File: src/installer.cpp [hunk 23]
```diff
@@ -25,8 +31,257 @@ using namespace flamingo;
...
+  // Middles
+  for (++it; std::next(it) != hooks.end(); ++it) {
+    it->assign_orig(std::next(it)->hook_ptr);
+  }
+
+  // Tail
+  // 'it' now refers to the last element
+  it->assign_orig(target_entry->second.metadata.metadata.need_orig
+                      ? target_entry->second.fixups.fixup_inst_destination.addr.data()
+                      : reinterpret_cast<void*>(&no_fixups));
```
Comment (sc2ad) [inline src/installer.cpp]: All of this logic can be written instead as (or as a for loop):

```
while (std::next(it) != hooks.end()) {
  it->assign_orig(std::next(it)->hook_ptr);
  ++it;
}
it->assign_orig(target_entry->second.metadata.metadata.need_orig
                        ? target_entry->second.fixups.fixup_inst_destination.addr.data()
                        : reinterpret_cast<void*>(&no_fixups));
```

File: src/installer.cpp [hunk 24]
```diff
@@ -35,20 +290,121 @@ Result<std::list<HookInfo>::iterator, installation::TargetBadPriorities> find_su
...
       // We cannot install here, we have a conflict
       return ResultT::Err(installation::TargetBadPriorities{
-        hook_to_install, fmt::format("Cannot install a 'final' hook after another 'final' hook with name: {}", hooks.back().metadata.name_info) });
+        hook_to_install.metadata, fmt::format("Cannot install a 'final' hook after another 'final' hook with name: {}",
+                                              hooks.back().metadata.name_info) });
     }
     // Select the end to install at
-    return ResultT::Ok(hooks.end());
+
+    auto new_it = hooks.emplace(hooks.end(), std::move(hook_to_install));
```
Comment (sc2ad) [inline src/installer.cpp]: nit: this could be `hooks.emplace_back` instead of emplace at end

File: src/installer.cpp [hunk 25]
```diff
@@ -35,20 +290,121 @@ Result<std::list<HookInfo>::iterator, installation::TargetBadPriorities> find_su
...
+    return ResultT::Ok(new_it);
+  }
+
+  // ok now we have a non-final hook
+  // if existing hooks have priority constraints that depend on us, we need to respect those
+  // therefore topological
+
+  // If the incoming hook has any priority constraints, we may need a topological pass.
+  bool requires_sort =
```
Comment (sc2ad) [inline src/installer.cpp:215]: I think if we can simplify the topo sort routine, we can avoid this checking logic entirely and just always do it. We will need to ensure we keep things stable, as you mentioned on discord, as I think that is an important feature that we should keep around, but otherwise doing the sort seems fine.

Recompile is cheap in comparison to a sort, so I wouldn't worry too much about that.

File: src/installer.cpp [hunk 26]
```diff
@@ -151,14 +515,15 @@ installation::Result Install(HookInfo&& hook) {
...
     return installation::Result::ErrAt<installation::TargetBadPriorities>(location_or_err.error());
   }
-  auto const location = location_or_err.value();
-  // 2. Assuming we found a reasonable location to install, insert our new hook before this location, and then adjust
-  // those around us to match.
-  auto const hook_data_result = hooked_target->second.hooks.emplace(location, std::move(hook));
+  auto const hook_data_result = location_or_err.value();
+
+  // TODO: Recompile fixups/origs for all hooks on this target or not?
+  // recompile_hooks(hooked_target->second.hooks, target_info);
```
Comment (sc2ad) [inline src/installer.cpp:459]: This recompile is currently in the "find" method, I'd prefer if we moved it out here and do it unconditionally after a found location (the recompile hooks method can decide if it wants to nop as an optimization, or not, in the futrure; doesn't need to do that now)

**Review by sc2ad — COMMENTED** [review]
Haven't read through all the sorting tests yet, but just left some high level comments here
